x-pack/plugins/security_solution/server/endpoint/services/actions/action_files.ts

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ElasticsearchClient, Logger } from '@kbn/core/server';
+import type { Readable } from 'stream';
+import { createEsFileClient } from '@kbn/files-plugin/server';
+import { errors } from '@elastic/elasticsearch';
+import { NotFoundError } from '../../errors';
+import {
+  FILE_STORAGE_DATA_INDEX,
+  FILE_STORAGE_METADATA_INDEX,
+} from '../../../../common/endpoint/constants';
+import { EndpointError } from '../../../../common/endpoint/errors';
+
+/**
+ * Returns a NodeJS `Readable` data stream to a file
+ * @param esClient
+ * @param logger
+ * @param fileId
+ */
+export const getFileDownloadStream = async (
+  esClient: ElasticsearchClient,
+  logger: Logger,
+  fileId: string
+): Promise<{ stream: Readable; fileName: string; mimeType?: string }> => {
+  const fileClient = createEsFileClient({
+    metadataIndex: FILE_STORAGE_METADATA_INDEX,
+    blobStorageIndex: FILE_STORAGE_DATA_INDEX,
+    elasticsearchClient: esClient,
+    logger,
+  });
+
+  try {
+    const file = await fileClient.get({ id: fileId });
+    const { name: fileName, mimeType } = file.data;
+
+    return {
+      stream: await file.downloadContent(),
+      fileName,
+      mimeType,
+    };
+  } catch (error) {
```

Comment:
do we know what error comes back if the metadata document exists, but no chunks exist? 

Should probably also be a 404? or 410. 

Although if this is a "fetched-too-soon" situation and it's still uploading, then 410 wouldn't be correct.


Post-deletion, that's a great status code. but 404 is ok for the unknown, and general purpose here.

Reply:
I'll test that out by deleting the chunks. I don't think we get an error here - I thin the error comes in when you start to actually read the chunks and they are not found.

the "fetched-too-soon" is not a real issue from our use case via Kibana UI, but could happen via API only if someone called it manually. if that was to occur, the Files service returns a `NoDownloadAvailableError` ([here](https://github.com/elastic/kibana/blob/01ecbd48f3657327cb5375b54195cd07bfc4027f/x-pack/plugins/files/server/file/file.ts#L122-L124)) which I think would translate into a `500` API error on our side